// google-calendar-parse/GoogleCalendarParse.js

exports.GoogleCalendarParse = GoogleCalendarParse;

var util   = require('util');
var querystring = require('querystring');

var OAuth = require('google-oauth');
var rest = require('restler');
var https = require('https');


function GoogleCalendarParse(parse_app, consumer_key, consumer_secret, callback_url){
  this.parse_app = parse_app
  this.key = consumer_key;
  this.cached_tokens = new Array();
  this.oauth = new OAuth.OAuth2(
    consumer_key, 
    consumer_secret,
    callback_url);
}

GoogleCalendarParse.prototype.refreshAccessToken = function(userId, callback) {
	var T = this;
	this.parse_app.find('User', userId, function(err, response) {
		if (!response) {
			callback(null);
			return;
		}
		var refresh_token = response.google_refresh_token;
		console.log("Using refresh token: ", refresh_token);
		var request_data = querystring.stringify({
					refresh_token: refresh_token,
					client_id: T.google_consumer_key,
					client_secret: T.google_consumer_secret,
					grant_type: 'refresh_token'
		});

		var options = {
			hostname: 'accounts.google.com',
			path: '/o/oauth2/token',
			method: 'POST',
			headers: {
				'Content-Type': 'application/x-www-form-urlencoded',
				'Content-Length': request_data.length
			}
		}

		var request = https.request(options, function(res) {
			console.log("GoogleParseAuth - statusCode: ", res.statusCode);
			console.log("GoogleParseAuth - headers: ", res.headers);
			res.setEncoding('utf8');
			res.on('data', function(data) {
				console.log("Got data response for refreshing token");
				data = JSON.parse(data);
				var new_access_token = data.access_token;
				T.parse_app.update('User', userId, {google_access_token: new_access_token}, function(err, response){});
				console.log("GoogleParseAuth - received new access token: " + new_access_token);
				callback(new_access_token);
			});
		});
		request.write(request_data);
		request.end();

		request.on('error', function(e) {
			console.error(e);
			callback(false);
		});
	});
}


GoogleCalendarParse.prototype.getGoogleAuthorizeTokenURL = function(callback) {
	return this.oauth.getGoogleAuthorizeTokenURL(['https://www.googleapis.com/auth/calendar'], callback)
}

GoogleCalendarParse.prototype.retrieveAccessToken = function(userId, done) {
    if (userId in this.cached_tokens) {
        done(false, cached_tokens[userId]);
    } else {
        this.parse_app.find('User', userId, function (err, response) {
            var access_token;
            if(response && response.google_access_token) {
                console.log("Using Parse auth token ", response.google_access_token);
                access_token = response.google_access_token;
                done(false, access_token);
            } else {
                console.log("Couldn't find access token");
                done(true, null);
            }
        });
    }
}

GoogleCalendarParse.prototype.sendRequest = function(type, hostname, path, user_id, option, body, callback, try_refresh) {
    var T = this;
    this.retrieveAccessToken(user_id, function(err, access_token) {
        if((callback === null || callback === undefined) && body !== null) {
            callback = body;
            body = null;
          }

          if((callback === null || callback === undefined) && option !== null) {
            callback = option;
            option = null;
          }

          if(body && typeof body == 'object'){
            body = JSON.stringify(body)
          } 
          
          callback = callback || function(){};
        
          option = option || {};
          option.access_token = access_token;
          option.key = T.key;
          
          var restRequest = null;
          var requestOption = { query:option, parser:rest.parsers.json };
          if(body){
            requestOption.data = body;
            requestOption.headers = {};
            requestOption.headers['content-type'] = 'application/json';
          }  

          var url = hostname + path;
          
          switch(type.toLowerCase()){
            
            case 'del':
            case 'delete': 
                restRequest = rest.del(url, requestOption);
              break;
              
            case 'put': restRequest = rest.put(url, requestOption);
              break;
            
            case 'post': restRequest = rest.post(url, requestOption);
              break;
            
            default : restRequest = rest.get(url, requestOption);
          }
          

          restRequest.on('complete', function(result, response ) {
            console.log("Completed rest request to ", url);
            console.log("Got result: ", result);
            if(result instanceof Error || response.statusCode != 200){
                if (try_refresh == null || try_refresh) {
                    console.log("Got error: ", result);
                    T.refreshAccessToken(user_id, function(access_token) {
                        T.sendRequest(type, url, user_id, option, body, callback, false);
                    });
                    return;
                } else {
                    return callback(result, response.rawEncoded);
                }
            }
            
            console.log("Done! Calling back now");
            return callback(null, result);
          });
        
        /*


            var request_data = querystring.stringify({
                    access_token: access_token,
                    key: T.google_consumer_key,
            });

            var options = {
                hostname: hostname,
                path: path,
                method: type,
                headers: {
                    'Content-Type': 'application/json',
                    'Content-Length': request_data.length
                }
            }

            var request = https.request(options, function(res) {
                console.log("GoogleCalendarParse - statusCode: ", res.statusCode);
                console.log("GoogleCalendarParse - headers: ", res.headers);
                res.setEncoding('utf8');
                res.on('data', function(data) {
                    console.log("Got data response for sending request: ", data);
                    if (res.statusCode != 200) {
                        return callback(true, null);
                    } else {
                        data = JSON.parse(data);
                        return callback(null, data);
                    }
                });
            });
            request.write(request_data);
            request.end();

            */

    });

  
}

// Calendar List

GoogleCalendarParse.prototype.listCalendarList = function(user_id, option, callback) {
  console.log("listCalendarList");
  return this.sendRequest('get','https://www.googleapis.com', '/calendar/v3/users/me/calendarList', user_id, option, callback)
} 

GoogleCalendarParse.prototype.getCalendarList = function(user_id, calendarId, callback) {
  console.log("getCalendarList");
  return this.sendRequest('get','https://www.googleapis.com', '/calendar/v3/users/me/calendarList/'+calendarId, user_id, option, callback)
}

// Events

GoogleCalendarParse.prototype.listEvent = function(user_id, calendarId, option, callback) {
  console.log("In GoogleCalendar, trying to listEvent");
  return this.sendRequest('get', 'https://www.googleapis.com', '/calendar/v3/calendars/'+encodeURIComponent(calendarId)+'/events', 
    user_id, option, callback);
}

GoogleCalendarParse.prototype.insertEvent = function(user_id, calendarId, event, option, callback) {
  
  if(arguments.length < 5){
    callback = option;
    option = {};
  }
  
  return this.sendRequest('post', 'https://www.googleapis.com', '/calendar/v3/calendars/'+encodeURIComponent(calendarId)+'/events', 
    user_id, option, event, callback);
}

GoogleCalendarParse.prototype.getEvent = function(user_id, calendarId, eventId, option, callback) {
  
  return this.sendRequest('get', 'https://www.googleapis.com', '/calendar/v3/calendars/'+encodeURIComponent(calendarId)+'/events/'+eventId, 
    user_id, option, callback);
}

GoogleCalendarParse.prototype.deleteEvent = function(user_id, calendarId, eventId, option, callback) {
  
  return this.sendRequest('delete', 'https://www.googleapis.com', '/calendar/v3/calendars/'+encodeURIComponent(calendarId)+'/events/'+eventId, 
    user_id, option, callback);
}

GoogleCalendarParse.prototype.updateEvent = function(user_id, calendarId, eventId, event, option, callback) {
  
  return this.sendRequest('put', 'https://www.googleapis.com', '/calendar/v3/calendars/'+encodeURIComponent(calendarId)+'/events/'+eventId, 
    user_id, option, event, callback);
}
